engine/src/main/resources/assets/shaders/chunk_frag.glsl

```diff
@@ -95,34 +101,29 @@ void main() {
 #if defined (NORMAL_MAPPING) || defined (PARALLAX_MAPPING)
     // TODO: Calculates the tangent frame on the fly - this is absurdly costly... But storing
     // the tangent for each vertex in the chunk VBO might be not the best idea either.
-    // Screen-space coordinates happen to be a coordinate system to which it's possible to relate both view-space and texture coordinates, but are themselves irrelevant.
+    // The specific relationship between screen coordinates and view coordinates is irrelevant here. Screen coordinates just happen to be a basis that it's possible to relate both view and UV space coordinates to.
     mat2x3 screenToView = mat2x3(dFdx(vertexViewPos.xyz), dFdy(vertexViewPos.xyz));
     mat2   screenToUv   = mat2  (dFdx(gl_TexCoord[0].xy), dFdy(gl_TexCoord[0].xy)) / TEXTURE_OFFSET;
-    //mat2 uvToScreem = inverse(screenToUv);
-    //inverse is not available in GLSL 1.20, so calculate it manually:
-        float det = screenToUv[0][0] * screenToUv[1][1] - screenToUv[1][0] * screenToUv[0][1];
-        mat2 uvToScreen = mat2(vec2(screenToUv[1][1],-screenToUv[0][1]),vec2(-screenToUv[1][0],screenToUv[0][0])) / det;
+    mat2 uvToScreen = inverse2(screenToUv);
     mat2x3 uvToView = screenToView * uvToScreen;
-    mat3 tbn = mat3(uvToView[0], uvToView[1], normal);
 
 #if defined (PARALLAX_MAPPING)
-    vec3 eyeTangentSpace = normalizedVPos * tbn;
+    vec2 viewDirectionTextureOffset = normalizedVPos * uvToView;
 
-    float height =  parallaxScale * texture2D(textureAtlasHeight, texCoord).r - parallaxBias;
-    //Ideally this should be divided by eyeTangentSpace.z, but in practice this looks better at
-    //low angles, as it can't extend beyond its original space.
-	texCoord += height * eyeTangentSpace.xy * TEXTURE_OFFSET;
+    float height = parallaxScale * texture2D(textureAtlasHeight, texCoord).r - parallaxBias;
+    //Ideally this should be divided by dot(normal, normalizedVPos), as the texture offset is the amount the light-ray travels in across the texture in the time it takes to traverse the distance "height" perpendicular to the surface, but in practice this way looks better at low angles, as the height-map can't make the triangle extend beyond its normal size.
+	texCoord += height * viewDirectionTextureOffset * TEXTURE_OFFSET;
```

Comment:
Hmmm... I find the concept/metaphor of "light-ray traveling" and "the time it takes" somewhat confusing, as in the context of rendering a single frame there usually is no time (except perhaps when motion blur is involved) and no moving light rays. =)

Is this typical Parallax Mapping terminology or is it a noble attempt to clarify an otherwise tricky concept? I mean, aren't you talking about two components of a vector? One along the normal from the surface and the other along the surface? Can it be rephrased in this light?

Reply:
I agree that referring to time here is a little inappropriate, I just couldn't think of a better way of phrasing it. I have no idea if there's standard parallax mapping terminology, but I assume that if someone knows the meaning of specific jargon for this technique, they don't need it explained anyway. What about `Ideally this should be divided by dot(normal, normalizedVPos), as the texture offset is the component parallel to the surface of a vector along the view direction whose coordinate perpendicular to the surface has value "height", but in practice this way looks better at low angles, as the height-map can't make the triangle extend beyond its normal size.`? Personally I find the metaphor of light physically travelling easier to understand than this version.

Reply:
> but I assume that if someone knows the meaning of specific jargon for this technique, they don't need it explained anyway.

True, but there are also people who will look at this code to learn rather than because they already know it and want to improve it. And then there are people that just need to understand it enough to fit it into a bigger context instead of wanting to know the specific technique  inside-out. Hence my push to make everything as clear and unambiguous as possible. I'm not writing this as a criticism in any way, it's just to describe where I come from and the other types of people that might go through this code. 

In regard to the wording, and starting from your latest example, let me counter propose:

_Ideally this should be divided by dot(normal, normalizedVPos), as TEXTURE_OFFSET is the component parallel to the surface of a vector along the view's forward axis, the other component being a vector perpendicular to the surface and having magnitude "height". In practice this way looks better at low angles, as the height-map can't make the triangle extend beyond its boundaries._

I'm hoping this keeps the meaning while breaking the long sentence a little and adding some further structure to it.

Reply:
What I'm calling "texure offset" is different from TEXTURE_OFFSET, which I suppose is unhelpful, and it's compounded by the fact that TEXTURE_OFFSET is a confusing name to start with.